### PR TITLE
8318130: SocksSocketImpl needlessly encodes hostname for IPv6 addresses

### PR DESCRIPTION
**Pull request by djelinski**
This patch fixes the encoding of IPv6 addresses in the URI provided to ProxySelector.

The original code percent-encoded the square brackets added around the IPv6 addresses, producing URIs that could not be parsed correctly.

The proposed code does not percent-encode the IPv6 addresses at all.

Added a test to verify that different forms of addresses are processed correctly. The test verifies the handling of IPv4 addresses, IPv6 without scope, IPv6 with numeric scope, IPv6 with textual scope (only if present on the machine running tests), hostnames and FQDNs

Tier1-3 tests passed.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8318130](https://bugs.openjdk.org/browse/JDK-8318130): SocksSocketImpl needlessly encodes hostname for IPv6 addresses (**Bug** - P4)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Aleksei Efimov](https://openjdk.org/census#aefimov) (@AlekseiEfimov - **Reviewer**)
 * [Michael McMahon](https://openjdk.org/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16265/head:pull/16265` \
`$ git checkout pull/16265`

Update a local copy of the PR: \
`$ git checkout pull/16265` \
`$ git pull https://git.openjdk.org/jdk.git pull/16265/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16265`

View PR using the GUI difftool: \
`$ git pr show -t 16265`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16265.diff">https://git.openjdk.org/jdk/pull/16265.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16265#issuecomment-1772152110)

### CODE REVIEW

**Comment by djelinski** [comment]
ProxySelector.getDefault may return null if someone explicitly calls ProxySelector.setDefault(null); otherwise it's initialized to an instance of DefaultProxySelector [here](https://github.com/openjdk/jdk/blob/4812cabaa489e99481facddce69686a9fee29c44/src/java.base/share/classes/java/net/ProxySelector.java#L73-L78).

The modified code is used for the common case. DefaultProxySelector reads system proxy settings every time.

**Comment by AlanBateman** [comment]
> ProxySelector.getDefault may return null if someone explicitly calls ProxySelector.setDefault(null); otherwise it's initialized to an instance of DefaultProxySelector

Right, I guess part of my puzzlement is to understand why ProxySelector loads sun.net.spi.DefaultProxySelector reflectively. Slightly off topic from this PR but it looks like this is left over from the past.

**Comment by djelinski** [comment]
> why ProxySelector loads sun.net.spi.DefaultProxySelector reflectively.

Yeah, I was curious about that too; as far as I can tell, this won't prevent a class loader deadlock, and I can't think of any other reason. I can file a PR to simplify that.

File: src/java.base/share/classes/java/net/SocksSocketImpl.java [hunk 1]
```diff
@@ -297,17 +299,18 @@ public ProxySelector run() {
             URI uri;
             // Use getHostString() to avoid reverse lookups
             String host = epoint.getHostString();
-            // IPv6 literal?
             if (epoint.getAddress() instanceof Inet6Address &&
-                (!host.startsWith("[")) && (host.indexOf(':') >= 0)) {
+                    (host.indexOf(':') >= 0) &&
+                    isIPv6LiteralAddress(host)) {
```
Comment (Michael-Mc-Mahon) [inline src/java.base/share/classes/java/net/SocksSocketImpl.java:304]: Is the check at line 303 still needed, now we are calling `isIPv6LiteralAddress` ?

Comment (djelinski) [reply src/java.base/share/classes/java/net/SocksSocketImpl.java:304]: it's optional, as is the check for `:`. I'll remove them.

File: src/java.base/share/classes/java/net/SocksSocketImpl.java [hunk 2]
```diff
@@ -297,17 +299,16 @@ public ProxySelector run() {
...
             String host = epoint.getHostString();
-            // IPv6 literal?
-            if (epoint.getAddress() instanceof Inet6Address &&
-                (!host.startsWith("[")) && (host.indexOf(':') >= 0)) {
+            if (isIPv6LiteralAddress(host)) {
                 host = "[" + host + "]";
+            } else {
+                host = ParseUtil.encodePath(host);
             }
+            // IPv6 literal?
```
Comment (jaikiran) [inline src/java.base/share/classes/java/net/SocksSocketImpl.java:307]: Hello Daniel, this comment seems to have been misplaced. I think it should have been a few lines above just when `isIPv6LiteralAddress` is being called.

**Review by jaikiran — APPROVED** [review]

**Comment by jaikiran** [comment]
I just noticed that the test isn't marked `othervm` but changes the JVM level `ProxySelector` in its test execution. I think we should mark it `othervm` to avoid interfering (or being interefered) by other tests.

File: src/java.base/share/classes/java/net/SocksSocketImpl.java (hunk 2 above)
Comment (djelinski) [reply src/java.base/share/classes/java/net/SocksSocketImpl.java:307]: Good catch, this comment doesn't make much sense here. `isIPv6LiteralAddress` is descriptive enough, I'll just remove the comment.

**Comment by djelinski** [comment]
Updated the test to run in othervm and removed the code to restore original ProxySelector.

**Review by jaikiran — APPROVED** [review]

**Review by AlekseiEfimov — APPROVED** [review]
The changes and the new test look good to me.

**Review by Michael-Mc-Mahon — APPROVED** [review]

**Pull request closed without merging**